### PR TITLE
gh-103590: do not wrap a single exception raised from a try-except*

### LINKED ISSUE #103590: [ExceptionGroup] Inconsistent raise inside except*
# Bug report
When raising inside except* block and the caught exception wasn't an ExceptionGroup originally, then raised exception doesn't get wrapped in ExceptionGroup:
```py
try:
    try:
        raise TypeError(1)  # prints ValueError(3)
        raise ExceptionGroup('', [TypeError(2)])  # prints ExceptionGroup('', [ValueError(3)])
    except* TypeError:
        raise ValueError(3)
except BaseException as e:
    print(repr(e))
```

# Your environment

- CPython versions tested on: 3.11.3
- Operating system and architecture: Windows 11 / Ubuntu 22.04

### Linked PRs
* gh-103665
* gh-104094
* gh-104095

### PR DESCRIPTION
**Pull request by iritkatriel**
Fixes #103590.

* Issue: gh-103590

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2023-04-21-16-12-41.gh-issue-103590.7DHDOE.rst
Comment (gvanrossum) [inline Misc/NEWS.d/next/Core and Builtins/2023-04-21-16-12-41.gh-issue-103590.7DHDOE.rst:1]: Can you add this to the what’s new too? Are we backporting this to 3.11? From the tests this looks fairly invasive.

**Review by gvanrossum — COMMENTED** [review]
Looking at the logic and the test changes I wonder if this is a good idea?

**Comment by iritkatriel** [comment]
> Looking at the logic and the test changes I wonder if this is a good idea?

The example on the issue looks pretty convincing: https://github.com/python/cpython/issues/103590#issue-1671335697

**Comment by gvanrossum** [comment]
> The example on the issue looks pretty convincing: [#103590 (comment)](https://github.com/python/cpython/issues/103590#issue-1671335697)

I find that comment a bit confusing, but a later [comment by Carl](https://github.com/python/cpython/issues/103590#issuecomment-1512247100) is indeed convincing.

LGTM.

**Review by gvanrossum — APPROVED** [review]

**Review by gvanrossum — APPROVED** [review]
+1

File: Doc/whatsnew/3.12.rst [hunk 1]
```diff
@@ -203,6 +203,11 @@ Other Language Changes
   wrapped by a :exc:`RuntimeError`. Context information is added to the
   exception as a :pep:`678` note. (Contributed by Irit Katriel in :gh:`77757`.)
 
+* When a ``try-except*`` construct handles the entire :exc:`ExceptionGroup` and
```
Comment (iritkatriel) [inline Doc/whatsnew/3.12.rst]:
```suggestion
* When a ``try-except*`` construct handles the entire :exc:`ExceptionGroup`
```

**Comment by gvanrossum** [comment]
Should this be backported to 3.11?

**Comment by iritkatriel** [comment]
> Should this be backported to 3.11?

I thought so, but then I got the impression that you and @ambv thought it shouldn't.

**Comment by gvanrossum** [comment]
So I'm changing my mind, there can't be too much code out there that would be broken by this fix. What does @ambv think?

**Comment by ambv** [comment]
This is a bug. I agree there is probably little code out there that will be affected by the difference.

However, it's somewhat of a subtle change so user code is unlikely to be tested against it, and if a difference is experienced, the error displayed might be surprising. I'm thinking: "it works on my local box but on CI there's an exception, both are 3.11". While we continue improving 3.11, some installations will be forever stuck at 3.11.1 or so.

But given 3.11's low adoption so far I tend to agree that backporting is probably harmless and ultimately more helpful.

Let's backport then. But let's also be explicit about the change in the docs then. 3.12 whatsnew suggests this change only made it to 3.12, we need to amend that with a mention like "also changed in 3.11.4", and mention this in the except-star docs for both releases.

**Comment by gvanrossum** [comment]
Good plan. Thanks!

**Pull request merged**